src/components/save-import.tsx

```diff
@@ -3,49 +3,41 @@ import Button from "./button";
 import sample from "../sample-save";
 
 interface Props {
-  setFilename: React.Dispatch<string>;
-  setSavedata: React.Dispatch<XMLDocument>;
+  onFilenameChange: React.Dispatch<string>;
+  onSavedataChange: React.Dispatch<XMLDocument>;
   style?: React.CSSProperties;
 }
 
-export default function SaveImport({ setFilename, setSavedata, style }: Props) {
+export default function SaveImport({
+  onFilenameChange,
+  onSavedataChange,
+  style,
+}: Props): React.JSX.Element {
   const inputRef = useRef<HTMLInputElement>(null);
 
-  const _import = (files: FileList | null) => {
+  const handleInputFileChange: React.ChangeEventHandler<HTMLInputElement> = (
+    event,
+  ) => {
+    const files = event.target.files;
     if (files === null) return;
     if (files.length === 0) return;
     const savefile = files[0];
-    setFilename(savefile.name);
-    getSavedata(savefile, (savedata: XMLDocument) => {
-      setSavedata(savedata);
-    });
-  };
 
-  const getSavedata = (savefile: File, callback: Function) => {
-    const reader = new FileReader();
-    reader.addEventListener("load", () => {
-      const parser = new DOMParser();
-      const savedataAsString = reader.result as string;
-      const savedata = parser.parseFromString(
-        savedataAsString,
-        "text/xml"
-      ) as XMLDocument;
-      callback(savedata);
-    });
-    reader.readAsText(savefile);
+    onFilenameChange(savefile.name);
+    getSavedata(savefile, onSavedataChange);
```

Comment:
`getSavedata` has no reference inside the `react` so it has been moved after the component to reduce its length.
I also provided `onSavedataChange` directly to `getSavedata` since the callback signature is the same.